app/models/work_packages/relations.rb

```diff
@@ -29,14 +29,23 @@ module WorkPackages::Relations
 
   included do
     # All relations of the work package in both directions.
-    # Mostly used to have the relations destroyed upon destruction of the work package.
     # rubocop:disable Rails/InverseOf
     has_many :relations,
              ->(work_package) {
                unscope(:where)
                  .of_work_package(work_package)
              },
-             dependent: :destroy
+             dependent: :destroy do
+      def visible(user = User.current)
+        # The work_package_focus_scope in here is used to improve performance by reducing the number of
+        # work packages that have to be checked for whether they are visible.
+        # Since the method looks for relations on the current work package, only work packages
+        # that are on either end of a relation with the current work package need to be considered.
+        # The number should be quite small compared to the total number of work packages.
+        merge(Relation.visible(user,
+                               work_package_focus_scope: Arel::Nodes::UnionAll.new(select(:from_id).arel, select(:to_id).arel)))
```

Comment:
May be better to use `Union` instead of `UnionAll`, as there will be duplicates

Reply:
Duplicates would not matter at this point. The `UNION` ends up in a subselect by id. And one could argue, weakly, that a union does not require to have the duplicates removed which will improve performance.